src/java/org/apache/cassandra/service/AbstractReadExecutor.java

```diff
@@ -157,6 +181,68 @@ public Row get() throws ReadFailureException, ReadTimeoutException, DigestMismat
         return handler.get();
     }
 
+    /**
+     * Compare difference between passed timestamp and start field against Threshold cutoffs. If the threshold is
+     * exceeded, the current p99 latency of the retry endpoint is added to the threshold as the "predicted performance"
+     */
+    public void writePredictedSpeculativeRetryPerformanceMetrics(long timestamp) {
+        InetAddress extraReplica = Iterables.getLast(targetReplicas);
+        for (Threshold threshold : Threshold.values()) {
+            writePredSpecRetryPerformanceMetricByThreshold(threshold, timestamp, extraReplica);
+        }
+    }
+
+    private void writePredSpecRetryPerformanceMetricByThreshold(Threshold threshold, long timestamp, InetAddress extraReplica) {
+        long thresholdTime;
+        TimeUnit unit;
+        switch (threshold) {
+            case SECONDS_5:
+                thresholdTime = 5;
+                unit = TimeUnit.SECONDS;
+                break;
+            case SECONDS_1:
+                thresholdTime = 1;
+                unit = TimeUnit.SECONDS;
+                break;
+            case MILLISECONDS_100:
+                thresholdTime = 100;
+                unit = TimeUnit.MILLISECONDS;
+                break;
+            case P50:
+                thresholdTime = (long) (cfs.metric.coordinatorReadLatency.getSnapshot().getMedian());
+                unit = TimeUnit.NANOSECONDS;
+                break;
+            case P95:
+                thresholdTime = (long) (cfs.metric.coordinatorReadLatency.getSnapshot().get95thPercentile());
+                unit = TimeUnit.NANOSECONDS;
+                break;
+            case P99:
+                thresholdTime = (long) (cfs.metric.coordinatorReadLatency.getSnapshot().get99thPercentile());
+                unit = TimeUnit.NANOSECONDS;
+                break;
+            default:
+                throw new IllegalStateException("Unexpected value: " + threshold);
+        }
+        if (thresholdTime < 1) {
+            // Don't want uninitialized percentile latencies to skew the metrics
+            return;
+        }
+        long extraReplicaP99Latency;
+        try {
+            extraReplicaP99Latency = DatabaseDescriptor.getEndpointSnitch().getP99Latency(extraReplica);
+        } catch (RuntimeException e) {
+            logger.error("Failed to get p99 latency from endpoint snitch to record predicted speculative retry metrics", e);
+            return;
+        }
+        thresholdTime = TimeUnit.NANOSECONDS.convert(thresholdTime, unit);
+        if (timestamp - start > thresholdTime && extraReplicaP99Latency > 0) {
```

Comment:
Another thing to consider here: if we can get the contact times for each replica, and we they both exceed the threshold, then retry would've been useless here.